daffodil-runtime1-unparser/src/main/scala/org/apache/daffodil/processors/unparsers/SuppressableSeparatorUnparser.scala

```diff
@@ -0,0 +1,268 @@
+      // we include the final afterDOS data output stream in the list
+      buf += afterDOS
+    }
+    Assert.invariant(maybeNext.isDefined) // we MUST find a path to the afterDOS.
+    val res = buf.toSeq
+    res
+  }
+
+  /**
+   * Once we know whether the length is zero/non-zero, then we decide to unparse the
+   * separator or not.
+   *
+   * If we're positional and potentially trailing, then this will only be Zero length
+   * if we're considering unparsing a trailing separator for an empty, with nothing following.
+   * So if ZL, no separator, otherwise we unparse the separator.
+   */
+  override def continuation(state: UState): Unit = {
+    import ZeroLengthStatus._
+    zlStatus_ match {
+      case Zero => {
+        // zero length, so we suppress the separator
+        //
+        // That means we do nothing. After the continuation, the underlying suspension
+        // system finalizes the data output streams.
+      }
+      case NonZero => {
+        // non-zero case. So we need the separator.
+        sepUnparser.unparse1(savedUstate)
+      }
+      case Unknown =>
+        Assert.invariantFailed("Should be known zero or non-zero by here.")
+    }
+  }
+}
+
+final class SuppressableSeparatorUnparser private (sepUnparser: Unparser,
+  override val context: TermRuntimeData,
+  override val suspendableOperation: SuspendableOperation)
+  extends PrimUnparser
+  with SuspendableUnparser {
+
+  override val childProcessors: Vector[Processor] = Vector(sepUnparser)
+
+  override def runtimeDependencies = Vector()
+}
+
+object SuppressableSeparatorUnparser {
+
+  def apply(sepUnparser: Unparser,
+    context: TermRuntimeData,
+    suspendableOperation: SuspendableOperation) = {
+    val res = new SuppressableSeparatorUnparser(sepUnparser, context, suspendableOperation)
+    Processor.initialize(res)
+    res
+  }
+
+}
+
+/**
+ * We need to isolate the regions that we have to test for ZL/notZL
+ * for separator suppression. So we want these regions to not share
+ * data output streams with anything outside them.
+ *
+ * To achieve that we split the data output stream using its buffering
+ * capabilities. Some of the splits are natural. They are one side of
+ * the (potentially suppressed) separator itself. So creating that
+ * suspendable unparser for the separator gives us a split.
+ *
+ * The other side of these regions is however, an artificial split.
+ *
+ * To avoid duplication of complex code paths, we create this
+ * suspendable unparser that exists purely to split the underlying DOS
+ * but maintain all the invariants of the underlying DOS code, i.e.,
+ * the streams still get finalized (and therefore collapsed) the regular
+ * way.
+ *
+ * We ARE depending here, on the fact that nothing collapses the
+ * DOS together while we're still in the process of trying to acertain
+ * the ZL/nonZL of these regions.
+ */
+final class SuspressionRegionSplitUnparser private (override val context: TermRuntimeData)
```

Comment:
Add performance comment about making these a special kind of suspension that doesn't save the state. Since these are just splits for the sake of isolating the DOS that need to be queried to measure zero/non-zero.